density_scaler/src/hex.rs

```diff
@@ -45,20 +45,29 @@ static HIP17_RES_CONFIG: [Option<HexResConfig>; 11] = [
     Some(HexResConfig::new(2, 1, 1)),     // 10
 ];
 
-#[derive(Debug, Serialize, Eq, PartialEq)]
-pub struct ScalingMap(HashMap<String, Decimal>);
+#[async_trait::async_trait]
+pub trait HexDensityMap: Clone {
+    async fn get(&self, hex: &str) -> Option<Decimal>;
+    async fn swap(&self, new_map: HashMap<String, Decimal>);
+}
+
+#[derive(Debug, Clone)]
+pub struct SharedHexDensityMap(Arc<RwLock<HashMap<String, Decimal>>>);
 
-impl ScalingMap {
+impl SharedHexDensityMap {
     pub fn new() -> Self {
-        Self(HashMap::new())
+        Self(Arc::new(RwLock::new(HashMap::new())))
     }
+}
 
-    pub fn insert(&mut self, hex: &str, scale_factor: Decimal) -> Option<Decimal> {
-        self.0.insert(hex.to_string(), scale_factor)
+#[async_trait::async_trait]
```

Comment:
do you actually need to annotate the trait _and_ the impl of the trait with `async_trait`? this seems like something the compiler should be able to figure out from the annotation on the trait itself

Reply:
I had to unless I am missing something.

Reply:
Rust doesn't currently support async trait functions. `async_trait` is a procedural macro that gets around this by turning all `async fn` into normal `fn` that return a future.